sys/net/gnrc/link_layer/lwmac/lwmac.c

```diff
@@ -0,0 +1,867 @@
+            random_backoff = random_uint32_range(0, LWMAC_TIME_BETWEEN_WR_US);
+            time_until_tx = time_until_tx + random_backoff;
+
+            lwmac_set_timeout(gnrc_netdev, TIMEOUT_WAIT_FOR_DEST_WAKEUP, time_until_tx);
+
+            /* Register neighbour to be the next */
+            gnrc_netdev->tx.current_neighbor = neighbour;
+
+            /* Stop dutycycling, we're preparing to send. This prevents the
+             * timeout arriving late, so that the destination phase would
+             * be missed. */
+            /* TODO: bad for power savings */
+            rtt_handler(LWMAC_EVENT_RTT_PAUSE, gnrc_netdev);
+        }
+    }
+    else {
+        if (lwmac_timeout_is_expired(gnrc_netdev, TIMEOUT_WAIT_FOR_DEST_WAKEUP)) {
+            LOG(LOG_DEBUG, "[lwmac] Got timeout for dest wakeup, ticks: %" PRIu32 "\n",
+                           rtt_get_counter());
+            gnrc_netdev_lwmac_set_tx_continue(gnrc_netdev, false);
+            gnrc_netdev->tx.tx_burst_count = 0;
+            lwmac_set_state(gnrc_netdev, LWMAC_TRANSMITTING);
+        }
+    }
+}
+
+static void _rx_management(gnrc_netdev_t *gnrc_netdev)
+{
+    lwmac_rx_state_t state_rx = gnrc_netdev->rx.state;
+
+    switch (state_rx) {
+        case RX_STATE_STOPPED: {
+            lwmac_rx_start(gnrc_netdev);
+            lwmac_rx_update(gnrc_netdev);
+            break;
+        }
+        case RX_STATE_FAILED: {
+            /* This may happen frequently because we'll receive WA from
+             * every node in range. */
+            LOG(LOG_DEBUG, "[lwmac] Reception was NOT successful\n");
+            lwmac_rx_stop(gnrc_netdev);
+
+            if (gnrc_netdev->rx.rx_bad_exten_count >= LWMAC_MAX_RX_EXTENSION_NUM) {
+                gnrc_netdev_lwmac_set_quit_rx(gnrc_netdev, true);
+            }
+
+            /* Here we check if we are close to the end of the cycle. If yes,
+             * go to sleep. Firstly, get the relative phase. */
+            uint32_t phase = rtt_get_counter();
+            if (phase < gnrc_netdev->lwmac.last_wakeup) {
+                phase = (RTT_US_TO_TICKS(LWMAC_PHASE_MAX) - gnrc_netdev->lwmac.last_wakeup) +
+                         phase;
+            }
+            else {
+                phase = phase - gnrc_netdev->lwmac.last_wakeup;
+            }
+            /* If the relative phase is beyond 4/5 cycle time, go to sleep. */
+            if (phase > (4*RTT_US_TO_TICKS(LWMAC_WAKEUP_INTERVAL_US)/5)) {
+                gnrc_netdev_lwmac_set_quit_rx(gnrc_netdev, true);
+            }
+
+            if (gnrc_netdev_lwmac_get_quit_rx(gnrc_netdev)) {
+                lwmac_set_state(gnrc_netdev, LWMAC_SLEEPING);
+            }
+            else {
+                /* Go back to LISTENING for keep hearing on the channel */
+                lwmac_set_state(gnrc_netdev, LWMAC_LISTENING);
+            }
+            break;
+        }
+        case RX_STATE_SUCCESSFUL: {
+            LOG(LOG_DEBUG, "[lwmac] Reception was successful\n");
```

Comment:
move to new function `static void _rx_management_success(gnrc_netdev_t *gnrc_netdev);`?